go/processor/floria/processor_test.go

```diff
@@ -116,41 +115,138 @@ func TestProcessor_BuyGasInsufficientBalance(t *testing.T) {
 }
 
 func TestGasUsed(t *testing.T) {
-	tests := []struct {
-		sender          tosca.Address
-		expectedGasUsed tosca.Gas
+	tests := map[string]struct {
+		transaction     tosca.Transaction
+		result          tosca.CallResult
+		revision        tosca.Revision
+		expectedGasLeft tosca.Gas
 	}{
-		{
-			sender:          tosca.Address{},
-			expectedGasUsed: 500,
+		"InternalTransaction": {
+			transaction: tosca.Transaction{
+				Sender:   tosca.Address{},
+				GasLimit: 1000,
+			},
+			result: tosca.CallResult{
+				GasLeft:   500,
+				Success:   true,
+				GasRefund: 0,
+			},
+			revision:        tosca.R10_London,
+			expectedGasLeft: 500,
 		},
-		{
-			sender:          tosca.Address{1},
-			expectedGasUsed: 550,
+		"NonInternalTransaction": {
+			transaction: tosca.Transaction{
+				Sender:   tosca.Address{1},
+				GasLimit: 1000,
+			},
+			result: tosca.CallResult{
+				GasLeft:   500,
+				Success:   true,
+				GasRefund: 0,
+			},
+			revision:        tosca.R10_London,
+			expectedGasLeft: 450,
+		},
+		"RefundPreLondon": {
+			transaction: tosca.Transaction{
+				Sender:   tosca.Address{},
+				GasLimit: 1000,
+			},
+			result: tosca.CallResult{
+				GasLeft:   500,
+				Success:   true,
+				GasRefund: 300,
+			},
+			revision:        tosca.R09_Berlin,
+			expectedGasLeft: 750,
 		},
-		{
-			sender:          tosca.Address{42},
-			expectedGasUsed: 550,
+		"RefundLondon": {
+			transaction: tosca.Transaction{
+				Sender:   tosca.Address{},
+				GasLimit: 1000,
+			},
+			result: tosca.CallResult{
+				GasLeft:   500,
+				Success:   true,
+				GasRefund: 300,
+			},
+			revision:        tosca.R10_London,
+			expectedGasLeft: 600,
 		},
-	}
-
-	for _, test := range tests {
-		t.Run(fmt.Sprintf("sender%v", test.sender), func(t *testing.T) {
-			transaction := tosca.Transaction{
-				Sender:   test.sender,
+		"RefundPostLondon": {
+			transaction: tosca.Transaction{
+				Sender:   tosca.Address{},
 				GasLimit: 1000,
-			}
+			},
+			result: tosca.CallResult{
+				GasLeft:   500,
+				Success:   true,
+				GasRefund: 300,
+			},
+			revision:        tosca.R13_Cancun,
+			expectedGasLeft: 600,
+		},
+		"smallRefund": {
+			transaction: tosca.Transaction{
+				Sender:   tosca.Address{},
+				GasLimit: 1000,
+			},
+			result: tosca.CallResult{
+				GasLeft:   500,
+				Success:   true,
+				GasRefund: 5,
+			},
+			revision:        tosca.R10_London,
+			expectedGasLeft: 505,
+		},
+		"UnsuccessfulResult": {
+			transaction: tosca.Transaction{
+				Sender:   tosca.Address{},
+				GasLimit: 1000,
+			},
+			result: tosca.CallResult{
+				GasLeft:   500,
+				Success:   false,
+				GasRefund: 500,
+			},
+			revision:        tosca.R10_London,
+			expectedGasLeft: 500,
```

Comment:
In case of an unsuccessful execution all gas should be consumed .. so this value should be `0`, right?

The implementation of `calculateGasLeft` seems to assume that in case of a unsuccessful execution `result.GasLeft` is `0`. This is probably the case for all interpreter implementations, and so the implementation will likely produce the correct result. If so, this test case describes an impossible input.

Reply:
Right, I mainly wanted to test that the refund is not added back in an unsuccessful case. I changed the gasLeft to 0 to avoid any confusion.